cli/tests/activate.bats

```diff
@@ -984,6 +984,69 @@ EOF
   assert_output --partial "test_alias is an alias for echo testing"
 }
 
+# bats test_tags=activate,activate:rc:zsh
+@test "zsh: interactive activate respects history settings from dotfile" {
+  project_setup
+
+  # This should always work, even when Darwin sets a default in `/etc/zshrc`.
+  echo 'HISTFILE=${PROJECT_DIR}/.alt_history' >"$HOME/.zshrc.extra"
+  echo 'SHELL_SESSION_DIR=${PROJECT_DIR}/.alt_sessions' >>"$HOME/.zshrc.extra"
+
+  # TODO: flox will set HOME if it doesn't match the home of the user with
+  # current euid. I'm not sure if we should change that, but for now just set
+  # USER to REAL_USER.
+  FLOX_SHELL="zsh" USER="$REAL_USER" NO_COLOR=1 \
+    run expect "$TESTS_DIR/activate/histfile.exp" "$PROJECT_DIR"
+  assert_success
+  assert_line --partial "HISTFILE=$PROJECT_DIR/.alt_history"
+  assert_line --partial "SHELL_SESSION_DIR=$PROJECT_DIR/.alt_sessions"
+}
+
+# bats test_tags=activate,activate:rc:zsh
+@test "zsh: interactive activate respects history settings from dotfile based on original ZDOTDIR" {
+  project_setup
+
+  # Mimic the default `/etc/zshrc` on Darwin prior to Nix being installed. We
+  # have to do this in `~/.zshrc` because we can't mock `/etc/zshrc`. However we
+  # apply the same `ZDOTDIR` logic to both, ensuring that it's not pointing at
+  # our immutable storepath.
+  echo 'HISTFILE=${ZDOTDIR:-$HOME}/.alt_history' >"$HOME/.zshrc.extra"
+  echo 'SHELL_SESSION_DIR=${ZDOTDIR:-$HOME}/.alt_sessions' >>"$HOME/.zshrc.extra"
+
+  # TODO: flox will set HOME if it doesn't match the home of the user with
+  # current euid. I'm not sure if we should change that, but for now just set
+  # USER to REAL_USER.
+  FLOX_SHELL="zsh" USER="$REAL_USER" NO_COLOR=1 \
+    run expect "$TESTS_DIR/activate/histfile.exp" "$PROJECT_DIR"
+  assert_success
+  assert_line --partial "HISTFILE=$HOME/.alt_history"
+  assert_line --partial "SHELL_SESSION_DIR=$HOME/.alt_sessions"
+}
+
+# bats test_tags=activate,activate:rc:zsh
+@test "zsh: interactive activate respects history settings from environment variable on Linux" {
+  project_setup
+
+  # TODO: flox will set HOME if it doesn't match the home of the user with
+  # current euid. I'm not sure if we should change that, but for now just set
+  # USER to REAL_USER.
+  FLOX_SHELL="zsh" USER="$REAL_USER" NO_COLOR=1 \
+    HISTFILE="$PROJECT_DIR/.alt_history" \
+    SHELL_SESSION_DIR="$PROJECT_DIR/.alt_sessions" \
+    run expect "$TESTS_DIR/activate/histfile.exp" "$PROJECT_DIR"
+
+  assert_success
+  assert_line --partial "SHELL_SESSION_DIR=$PROJECT_DIR/.alt_sessions"
+
+  # This is only honoured when `/etc/zshrc` and `~/.zshrc` don't set a default
+  # like Darwin does.
+  if [[ "$(uname)" == "Darwin" ]]; then
+    assert_line --partial "HISTFILE=$HOME/.zsh_history"
+  else
+    assert_line --partial "HISTFILE=$PROJECT_DIR/.alt_history"
+  fi
```

Comment:
```suggestion
  refute_line --partial "HISTFILE=/nix/store/"
  refute_line --partial "SHELL_SESSION_DIR=/nix/store/"
```

We are not in control of the `/etc/zshrc` file on any system, so don't attempt to predict whether the HISTFILE value has or has not been overridden (as it is by default on Apple). Instead let's just assert that we don't end up setting either of `HISTFILE` or `SHELL_SESSION_DIR` to an immutable `/nix/store` path.

See comment above for why this test otherwise fails for me on my host where I have fixed the Apple `/etc/zshrc` bug.

Reply:
FTR I also have this patch:
```
flox [tmp.HWJY0RfQuo limeytexan/default (remote)] brantley@Michaels-MBP tmp.HWJY0RfQuo % diff /etc/zshrc_Apple_Terminal{.backup-before-flox,}                                                                           
102c102
<     SHELL_SESSION_DIR="${ZDOTDIR:-$HOME}/.zsh_sessions"
---
>     SHELL_SESSION_DIR="${SHELL_SESSION_DIR:-${ZDOTDIR:-$HOME}/.zsh_sessions}"
```

Reply:
> Because it's broken, and because anyone in the world could similarly fix the bug in the way I did. Or, some other Linux distro could make the same mistake Apple did and then the Darwin conditional would be wrong.

Can you elaborate on why it's broken, beyond the context of Flox previously passing the wrong `ZDOTDIR`?

> Again, the bad thing that happened originally is that we set the ZDOTDIR to a /nix/store path, then Apple's /etc/zshrc picked that up and used it to set HISTFILE and SHELL_SESSION_DIR, and then the user encountered errors on account of both of those being immutable. That's the problem we're seeking to detect in our tests, and we shouldn't attempt to predict what any given distro will do in their /etc/zshrc.

That's one of the bugs. Then the second bug that we're fixing here is that Flox unconditionally set a value after the user's session and dotfiles. My intention with the test is that I want to avoid regressing the second bug independently of the first bug and having to learn about the oddities of Linux/Mac default behaviour all over again.

Reply:
What makes me particularly nervous about negative-only assertions is that they will continue to pass and give us false confidence if all other parts of the test are modified, e.g. we typo or stop passing the environment variables to `expect`, change the output format of the `expect` script, or possibly even if we don't prefix the command with `run`.

How about this as a compromise?
```sh
  # If the host system honours the environment variables then we do too. The
  # majority of Linux distros don't ship with a default `/etc/zshrc` and will
  # match the "true" condition. Darwin, with or without Nix, ships with a
  # default `/etc/zshrc` that matches the "false" condition.
  if [[ "$(HISTFILE=/dev/null zsh -ic 'echo $HISTFILE')" == "/dev/null" ]]; then
    assert_line --partial "HISTFILE=$PROJECT_DIR/.alt_history"
  else
    assert_line --partial "HISTFILE=$HOME/.zsh_history"
  fi

  if [[ "$(SHELL_SESSION_DIR=/dev/null zsh -ic 'echo $SHELL_SESSION_DIR')" == "/dev/null" ]]; then
    assert_line --partial "SHELL_SESSION_DIR=$PROJECT_DIR/.alt_sessions"
  else
    assert_line --partial "SHELL_SESSION_DIR=$HOME/.zsh_sessions"
  fi
```

Reply:
@limeytexan I've pushed the above revision. Could you re-review the approach and try locally?

Reply:
> Can you elaborate on why it's broken, beyond the context of Flox previously passing the wrong ZDOTDIR?

It's broken because no user can define their own custom `HISTFILE` in either of their `~/.zshenv` or `~/.zshenv` files and not have it overridden by the system default.